File: js/queoperadora.js
/*
	queOperadora
	Copyright (c) 2012 Eduardo Orige (www.eduardoorige.com.br)
	Licensed under the MIT license (http://www.eduardoorige.com.br/projetos/queoperadora) 
	Version: 0.1
	
	Operadoras Disponiveis: 
		- Amazonia Celular 		-> Oi
		- Brasil Telecom 		-> Oi
		- BrT 					-> Oi
		- Claro
		- CTBC
		- Oi
		- Sercomtel
		- Telemig 				-> Vivo
		- TIM
		- Unicel 				-> Não existe mais
		- Vivo

*/
function queOperadora(ddd,fone){
	cel = (fone.charAt(0) + fone.charAt(1));
	pre_cel = (fone.charAt(0) + fone.charAt(1) + fone.charAt(2) + fone.charAt(3));

	if (ddd >= 11 && ddd <= 19)	{
		if (pre_cel >= 6340 && pre_cel <= 6369)		{
			return "TIM";
		}
		else {
			if ((pre_cel >= 6310 && pre_cel <= 6339) || (pre_cel >= 6589 && pre_cel <= 6599) ||
				(pre_cel >= 7052 && pre_cel <= 7062) ||	cel == 76 ||
				(pre_cel >= 8800 && pre_cel <= 8899) ||	cel == 89)	{
					return "CLARO";
			} else {
				if ((pre_cel >= 6057 && pre_cel <= 6060) || (pre_cel >= 6193 && pre_cel <= 6199) ||
					(pre_cel >= 6370 && pre_cel <= 6499) || pre_cel == 7099 ||
					(cel >= 71 && cel <= 75) || cel == 95) {
						return "VIVO";
				} else {
					if ((pre_cel >= 6100 && pre_cel <= 6193) || pre_cel == 6299 || cel == 65 || 
						(pre_cel >= 6651 && pre_cel <= 6799) || cel == 68 || pre_cel == 6999 ||
						(pre_cel >= 7971 && pre_cel <= 7999) || cel == 80 ){
						//(pre_cel >= 8814 && pre_cel <= 8899)) {
							return "OI";
					} else {
						if (cel >= 96 && cel <= 99) {
							return "VIVO";
						} else {
							if (cel >= 91 && cel <= 94){
								return "CLARO";
							} else {
								if (cel >= 81 && cel <= 87){
									return "TIM";
								} //	elif int(cel[0]) == 8 and int(cel[1]) != 1 or int(cel[1]) != 7:
								else {
									return "";
								}
							}
						}
					}
				}
			}
		}
	} else { //ACABA SAOPAULO
		if (ddd >= 21 && ddd <= 28){ //RJ, ES
			if (cel >= 74 && cel <= 76){
				return "CLARO";
			} else {
				if (cel == 71 || cel == 72 || cel == 95){
					return "VIVO";
				} else {
					if ((cel >= 84 && cel <= 86) || (cel >= 86 && cel <= 89)){
						return "OI";
					} else {
						if (cel >= 96 && cel <= 99){
							return "VIVO";
						} else {
							if (cel >= 91 && cel <= 94){
								return "CLARO";
							} else {
								if (cel >= 80 && cel <= 83) {
									return "TIM";
								} else {
									return "";
								}
							}
						}
					}
				}
			}
		} else { //ACABA RJ, ES
			if (ddd >= 31 && ddd <= 38) { //MG
				if (cel == 85 || cel == 86 || cel == 89){
					return "OI";
				} else {
					//TODO if (cel == 86)
					if ((pre_cel >= 9960 && pre_cel <= 9979) || (pre_cel >= 9991 && pre_cel <= 9999)){
						return "CTBC";
					} else {
						if (cel >= 96 && cel <= 99){
							return "VIVO";
						} else {
							if (cel >= 91 && cel <= 94){
								return "TIM";
							} else {
								if (cel == 87 || cel == 88){
									return "OI";										
								} else {
									if (cel >= 81 && cel <= 84){
										return "CLARO";
									} else {
										return "";
									}
								}
							}
						}
					}
				}
			} else { //ACABA MG
				if (ddd >= 41 && ddd <= 49) { //SC,PR
					if (ddd == 43 && pre_cel == 9998){
						return "TIM";
					} else {
						//TODO
						//Londrina e Tamarana, PR
						//if (pre_cel >= 9941 && pre_cel <= 9998){ window.alert("SERCOMTEL");} else {
						if (cel >= 96 && cel <= 99 ){
							return "TIM";
						} else {
							if (cel >= 91 && cel <= 94) {
								return "VIVO";
							} else {
								if (cel == 87 || cel == 88) {
									return "CLARO";
								} else {
									if (cel == 84 || cel == 85) {
										return "OI";
									} else {
										return "";
									}
								}
							}
						}
					}
				} else { //ACABA SC,PR
					if (ddd >= 51 && ddd <= 55){ //RS
						//TODO
						//if (pre_cel >= 9911 <= pre_cel 9939){ window.alert("TIM")}
						if (cel == 81 || cel == 82) {
							return "TIM";
						} else {
							if (cel == 84 || cel == 85){
								return "OI";
							} else {
								if (cel >= 91 && cel <= 84){
									return "CLARO";
								} else {
									if (cel >= 95 && cel <= 99) {
										return "VIVO";
									} else {
										return "";
									}
								}
							}
						}
					} else {//ACABA RS
						if (ddd >= 61 && ddd <= 69) { //CENTRO-OESTE
							if (pre_cel >= 9551 && pre_cel <= 9959){
								return "CLARO";
							} else {
								if (cel == 86){
									return "OI";
								} else {
									//TODO
									// if cel == 99 e nao existe vivo {return "CTBC";}
									if (cel >= 96 && cel <= 99){
										return "VIVO";
									} else {
										if (cel >= 91 && cel <= 94) {
											return "CLARO";
										} else {
											if (cel == 81 || cel == 82){
												return "TIM";
											} else {
												if (cel == 84 || cel == 85){
													return "OI";
												} else {
													return "";
												}
											}
										}
									}
								}
							}
						} else { //ACABA CENTRO-OESTE
							if (ddd >= 71 && ddd <= 79) { //BA,SE
								if (cel >= 81 && cel <= 83){
									return "CLARO";
								} else {
									if (cel >= 86 && cel <= 88){
										return "OI";
									} else {
										if (cel >= 91 && cel <= 94) {
											return "TIM";
										} else {
											if (cel >= 96 && cel <= 99) {
												return "VIVO";
											} else {
												return "";
											}
										}
									}
								}
							} else { //ACABA BA,SE
								if (ddd >= 81 && ddd <= 89){ //NORDESTE
									if ((pre_cel >= 8719 && pre_cel <= 8721) || (cel == 88)){
										return "TIM";
									} else {
										if ((pre_cel >= 8100 && pre_cel <= 8200) || (cel >= 81 && cel <= 83)){
											return "VIVO";
										} else {
											if (cel >= 84 && cel <= 86) {
												return "OI";
											} else {
												if (cel >= 96 && cel <= 99) {
													return "TIM";
												} else {
													if (cel >= 91 && cel <= 94) {
														return "CLARO";
													} else {
														return "";
													}
												}
											}
										}
									}
								} else {// ACABA NORDESTE
									if (ddd >= 91 && ddd <= 99) { //NORTE
										if (cel >= 80 && cel <= 83) {
											return "OI";
										} else {
											if (cel == 84){
												return "CLARO";
											} else {
												if (cel == 87 || cel == 88) {
													return "OI";
												} else {
													if (cel >= 91 && cel <= 94) {
														return "VIVO";
													} else {
														if (cel >= 96 && cel <= 99) {
															return "OI";
														} else {
															return "";
														}
													}
												}
											}
										}
									}
								}
							}
						} 
					}
				}	
			} 
		} 
	}
}

File: python/queoperadora.py
#-*- coding: utf-8 -*-
__author__ = "Eduardo Orige"
__credits__ = ["Eduardo Orige"]
__version__ = "0.02"
__maintainer__ = "Eduardo Orige"
__email__ = "eduardo.orige@gmail.com"
__status__ = "Em Producao"

import sys

class QueOperadora:
	"""
	Classe com todos os métodos utilizados para verificar um número de celular e sua operadora.
	"""
	def __init__(self):
		pass
		
	def operadoras(self):
		"""
		Mostra as operadoras que estão disponiveis para analise no sistema.
		"""
		operadoras_disponiveis = ['Claro','CTBC','Oi','Sercomtel','TIM','Vivo',]
									
		return sorted(operadoras_disponiveis)

	def regexCelular(self,celular):
		"""
		Função que verifica se o celular passado corresponde ao padrão solicitado
		e se o número tambem é valido.
		Ex: 99 9999-9999
		"""
		import re
		test = re.compile('([1-9][1-9]\s[6-9]\d{3}\-\d{4}$)')
		if test.match(celular):
			return True
		else:
			return False

	def queOperadora(self,celular):
		"""		
		Verifica corretamente seguindo o padrao anterior de numero de celular: 99 9999-9999
		Se padrão nao for seguido, função retorna False.
		
		Primeiro verifica-se as exceções dos prefixos com 4 digitos depois então verifica-se 
		os numeros "padroes".
		"""		
		ddd = int(celular[:2])
		cel = int(celular[3:5])
		pre_cel = int(celular[3:7])
		#---- SP -----#
		if 11 <= ddd <= 19:
			if 6340 <= pre_cel <= 6369:
				return "TIM"
			
			elif (	6300 <= pre_cel <= 6339 or
					6168 <= pre_cel <= 6181 or
					6589 <= pre_cel <= 6599 or
					7052 <= pre_cel <= 7062 or
					cel == 76 			or
					8800 <= pre_cel <= 8899 ):
					
				return "CLARO"
			
			elif ( 	6057 <= pre_cel <= 6060 or
					6182 <= pre_cel <= 6199 or
					6370 <= pre_cel <= 6499 or
					pre_cel == 7099 or
					71 <= cel <= 75 or
					cel == 95 or
					cel == 99 ):
				return "VIVO"
				#TODO
				#
				#TESTADO E NÃO ENCONTRADOS NUMEROS SUFICIENTES DA CTBC
				#elif cel == 99 onde a vivo nao esta: return CTBC
			
			elif (	6100 <= pre_cel <= 6167	or
					6200 <= pre_cel <= 6299 or
					cel == 65 or
					6651 <= pre_cel <= 6799 or
					cel == 68 or
					pre_cel == 6999 or
					7971 <= pre_cel <= 7999 or
					8010 <= pre_cel <= 8099):
				return "OI"
				
			else:	
				if 81 <= cel <= 87:
					return "TIM"

				elif 91 <= cel <= 94:
					return "CLARO"

				elif 96 <= cel <= 99:
					return "VIVO"

				else:
					return None

		#---- RJ, ES ----#
		elif ddd == 21 or ddd == 22 or ddd == 24 or ddd == 27 or ddd == 28:
			if cel == 71 or cel == 72 or cel == 95:
				return "VIVO"
			
			elif 74 <= cel <= 76:
				return "CLARO"
			
			elif 80 <= cel <= 83:
				return "TIM"
			
			elif 84 <= cel <= 86 or 87 <= cel <= 89:
				return "OI"

			elif 91 <= cel <= 94:
				return "CLARO"
							
			elif 96 <= cel <= 99:
				return "VIVO"
				
			else:
				return None

		#---- MG ----#
		elif  31 <= ddd <= 35 or ddd == 37 or ddd == 38:
			if 81 <= cel <= 84:
				return "CLARO"
			
			elif 85 <= cel <= 89:
				return "OI"

			elif 91 <= cel <= 94:
				return "TIM"

			elif 95 <= cel <= 99:
				return "VIVO"
			
			#TESTADO E NAO ENCONTRADO CELULARES COM CTBC
			#elif ( 	9960 <= pre_cel <= 9979 or 
			#		9991 <= pre_cel <= 9999 ):
			#	return "CTBC"

			else:
				return None	

		#---- SC, PR ----#
		elif 41 <= ddd <= 49:
			if ddd == 43 and cel == 81:
				return "TIM"

			elif ddd == 43 and 9941 <= pre_cel <= 9998:
				return "SERCOMTEL"

			else:
				if 84 <= cel <= 85:
					return "OI"

				elif 87 <= cel <= 88:
					return "CLARO"

				elif 91 <= cel <= 94:
					return "VIVO"
					
				elif 96 <= cel <= 99:
					return "TIM"

				else:
					return None

		#---------- RS --------------#
		elif ddd == 51 or 53 <= ddd <= 55:
			if ddd == 53 and 9911 <= pre_cel <= 9920: #PELOTAS E REGIAO
				return "TIM"
			else :	
				if 81 <= cel <= 82:
					return "TIM"

				elif 84 <= cel <= 85:
					return "OI"
				
				elif 91 <= cel <= 94:
					return "CLARO"
			
				elif 95 <= cel <= 99:
					return "VIVO"
			
				else:
					return None	 

		#---- Centro Oeste ---------#
		elif 61 <= ddd <= 69:
			if 9551 <= pre_cel <= 9559:
				return "CLARO"
			
			elif cel == 86:
				return "OI"
			
			#TESTES EFETUADOS E NUMERO DE CELULARES CTBC É MUITO BAIXO
			#elif cel == 99 : onde nao existe vivo
			#return "CTBC"
			
			else:
				if 81 <= cel <= 82:
					return "TIM"

				elif 84 <= cel <= 85:
					return "OI"

				elif 91 <= cel <= 94:
					return "CLARO"

				elif 96 <= cel <= 99:
					return "VIVO"

				else:
					return None	

		#---------- BA, SE ----------#
		elif ddd == 71 or ddd == 73 or ddd == 74 or ddd == 75 or ddd == 77 or ddd == 79:
			if 81 <= cel <= 83:
				return "CLARO"

			elif 86 <= cel <= 88:
				return "OI"
			
			elif 91 <= cel <= 94:
				return "TIM"
					
			elif 96 <= cel <= 99:
				return "VIVO"

			else:
				return None
				
		#---------- Nordeste --------------#
		elif 81 <= ddd <= 89:
			if 84 <= cel <= 86:
				return "OI"
			
			elif 8719 <= pre_cel <= 8721:
				return "TIM"
				
 			else:
				if 81 <= cel <= 83:
					return "VIVO"

				elif 87 <= cel <= 88:
					return "OI"

				elif 91 <= cel <= 94:
					return "CLARO"

				elif 96 <= cel <= 99:
					return "TIM"

				else:
					return None

		#---------- Norte --------------#
		elif 91 <= ddd <= 99:
			if 80 <= cel <= 83:
				return "OI"

			elif cel == 84:
				return "CLARO"

			elif 87 <= cel <= 88:
				return "OI"

			elif 91 <= cel <= 94:
				return "VIVO"

			elif 96 <= cel <= 99:
				return "OI"

			else:
				return None
		else:
			return False

	def terminal(self):

		Error = "Utilize: \npython queoperadora.py [DDD] [CELULAR]\npython queoperadora.py 99 9999-9999"
		

		if len(sys.argv) == 3:
			
			celular = ""
			for param in sys.argv[1:]: celular += param + " "

			
			operadora = self.queOperadora(celular)

			if operadora == None: 
				print "! Operadora não encontrada !"
			else: 
				print operadora
			
			sys.exit(0)
		
		else:
			print Error
			sys.exit(1)
		
	def main(self):
		self.terminal()



if __name__ == '__main__':
	q = QueOperadora()
	q.terminal()